src/main/java/org/kaazing/netx/ws/WebSocket.java

```diff
@@ -0,0 +1,247 @@
+ * messaging via the Kaazing Gateway.
+ * <p>
+ * Refer to {@link http://www.w3.org/TR/websockets/} for the published standard
+ * W3C WebSocket API specification.
+ * <p>
+ * Instances of {@link WebSocket} can be created using
+ * {@link WebSocketFactory#createWebSocket(java.net.URI, String...)} API.
+ */
+public abstract class WebSocket implements Closeable {
+    /**
+     * Adds the specified {@link WebSocketExtension} object to the list of enabled extensions. The extension's
+     * <code>toString()</code> method must return RFC-3864 formatted string so that it can be negotiated using the
+     * <i>Sec-Websocket-Extensions</i> header during the opening handshake.
+     *
+     * @param extension {@link WebSocketExtension} with <code>toString()</code> method that returns RFC-3864 formatted string
+     */
+    public abstract void addEnabledExtension(WebSocketExtension extension);
+
+    /**
+     * Disconnects with the server. This is a blocking call that returns only when the shutdown is complete.
+     *
+     * @throws IOException    if the disconnect did not succeed
+     */
+    @Override
+    public abstract void close() throws IOException;
+
+    /**
+     * Disconnects with the server with code. This is a blocking call that returns only when the shutdown is complete. An
+     * IllegalArgumentException is thrown if the code isn't 1000 or out of 3000 - 4999 range.
+     *
+     * @param code                         the error code for closing
+     * @throws IOException                 if the disconnect did not succeed
+     */
+    public abstract void close(int code) throws IOException;
+
+    /**
+     * Disconnects with the server with code and reason. This is a blocking call that returns only when the shutdown is complete.
+     *
+     * @param code                         the error code for closing
+     * @param reason                       the reason for closing
+     * @throws IOException                 if the disconnect did not succeed OR if the code isn't 1000 or out of range
+     *                                     3000 - 4999 OR if the reason is more than 123 bytes
+     */
+    public abstract void close(int code, String reason) throws IOException;
+
+    /**
+     * Connects with the server using an end-point. This is a blocking call. The thread invoking this method will be blocked
+     * till a successful connection is established. If the request is not upgraded as per RFC-6455, then IOException is thrown.
+     *
+     * @throws IOException  if the connection cannot be established or the upgrade is not successful
+     */
+    public abstract void connect() throws IOException;
+
+    /**
+     * Gets the {@link ChallengeHandler} that is used during authentication both at the connect-time as well as at subsequent
+     * revalidation-time that occurs at regular intervals.
+     *
+     * @return ChallengeHandler
+     */
+    public abstract ChallengeHandler getChallengeHandler();
```

Comment:
Do you think it would be reasonable to keep the `ChallengeHandler` to `WebSocketFactory` only and remove it from here?

There doesn't appear to be a valid usecase for `getChallengeHandler` in the public API for the application since it's just used by the `WebSocket` internals.